sway-core/src/ir_generation/function.rs

```diff
@@ -318,41 +319,65 @@ impl FnCompiler {
     fn compile_intrinsic_function(
         &mut self,
         context: &mut Context,
-        kind: TypedIntrinsicFunctionKind,
+        TypedIntrinsicFunctionKind {
+            kind,
+            arguments,
+            type_arguments,
+            span: _,
+        }: TypedIntrinsicFunctionKind,
         span: Span,
     ) -> Result<Value, CompileError> {
         match kind {
-            TypedIntrinsicFunctionKind::SizeOfVal { exp } => {
+            Intrinsic::SizeOfVal => {
+                // We index the array with confidence that the type-checker did its job.
+                let exp = arguments[0].clone();
                 // Compile the expression in case of side-effects but ignore its value.
                 let ir_type = convert_resolved_typeid(context, &exp.return_type, &exp.span)?;
-                self.compile_expression(context, *exp)?;
+                self.compile_expression(context, exp)?;
                 Ok(Constant::get_uint(
                     context,
                     64,
                     ir_type_size_in_bytes(context, &ir_type),
                     None,
                 ))
             }
-            TypedIntrinsicFunctionKind::SizeOfType { type_id, type_span } => {
-                let ir_type = convert_resolved_typeid(context, &type_id, &type_span)?;
+            Intrinsic::SizeOfType => {
+                // We index the array with confidence that the type-checker did its job.
+                let targ = type_arguments[0].clone();
+                let ir_type = convert_resolved_typeid(context, &targ.type_id, &targ.span)?;
                 Ok(Constant::get_uint(
                     context,
                     64,
                     ir_type_size_in_bytes(context, &ir_type),
                     None,
                 ))
             }
-            TypedIntrinsicFunctionKind::IsRefType { type_id, type_span } => {
-                let ir_type = convert_resolved_typeid(context, &type_id, &type_span)?;
+            Intrinsic::IsReferenceType => {
+                // We index the array with confidence that the type-checker did its job.
+                let targ = type_arguments[0].clone();
+                let ir_type = convert_resolved_typeid(context, &targ.type_id, &targ.span)?;
                 Ok(Constant::get_bool(context, !ir_type.is_copy_type(), None))
             }
-            TypedIntrinsicFunctionKind::GetStorageKey => {
+            Intrinsic::GetStorageKey => {
                 let span_md_idx = MetadataIndex::from_span(context, &span);
                 Ok(self
                     .current_block
                     .ins(context)
                     .get_storage_key(span_md_idx, None))
             }
+            Intrinsic::Eq => {
+                // We index the array with confidence that the type-checker did its job.
+                let lhs = arguments[0].clone();
+                let rhs = arguments[1].clone();
```

Comment:
One thing we _could_ do is codify the arguments into the enum itself, so we don't have a panic risk here.

e.g.

```rust
enum Intrinsic {
  Eq { lhs: TypedExpression, rhs: TypedExpression },
  SizeOf { type_argument: TypeInfo } 
}
```

etc...

I'm just suggesting this, but I wouldn't block this PR on it, as it could be a more major refactor. It would give us type safety here, though.